topic_dynamics/parsing.py

```diff
@@ -0,0 +1,376 @@
+
+from tqdm import tqdm
+import tree_sitter
+
+from .parsers.utils import get_parser
+from .slicing import get_dates, checkout_by_date
+
+NODE_TYPES = {"c": ["identifier", "type_identifier"],
+              "c-sharp": ["identifier", "type_identifier"],
+              "cpp": ["identifier", "type_identifier"],
+              "java": ["identifier", "type_identifier"],
+              "python": ["identifier", "type_identifier"]}
+
+
+class DataForWriting:
+    def __init__(self, a, b):
+        self.a = a, self.b = b
+
+    def __repr__(self):
+        return str(self.a) + "," + str(self.b)
+
+    @classmethod
+    def parse(string):
+        a, b = map(int, string.split(","))
+        return DataForWriting(a, b)
+
+
+def get_extensions(lang: str) -> str:
+    """
+    Returns the extension for a given language. TODO: get rid of this and add enry.
+    :param lang: language name.
+    :return: the extension.
+    """
+    extensions = {"cpp": "cpp",
+                  "java": "java",
+                  "python": "py"}
+    return extensions[lang]
+
+
+def get_files(directory: str, extension: str) -> List[str]:
+    """
+    Get a list of files with a given extension.
+    :param directory: the root directory that is studied.
+    :param extension: extension of the listed files.
+    :return: list of file paths.
+    """
+    dir_path = Path(directory)
+    files = [str(path) for path in sorted(dir_path.rglob("*." + extension))]
+    return files
+
+
+def read_file(file: str) -> bytes:
+    """
+    Read the contents of the file.
+    :param file: address of the file.
+    :return: bytes with the contents of the file.
+    """
+    with open(file, "r") as fin:
+        code = bytes(fin.read(), "utf-8")
+    return code
+
+
+def get_positional_bytes(node: tree_sitter.Node) -> Tuple[int, int]:
+    """
+    Extract start and end byte.
+    :param node: node on the AST.
+    :return: (start byte, end byte).
+    """
+    start = node.start_byte
+    end = node.end_byte
+    return start, end
+
+
+def get_identifiers(file: str, lang: str) -> List[Tuple[str, int]]:
+    """
+    Gather a sorted list of identifiers in the file and their count.
+    :param file: address of the file.
+    :param lang: the language of file.
+    :return: a list of tuples, identifier and count.
+    """
+    code = read_file(file)
+    tree = get_parser(lang).parse(code)
+    root = tree.root_node
+    identifiers = []
+
+    def traverse_tree(node: tree_sitter.Node) -> None:
+        """
+        Run down the AST (DFS) from a given node and gather identifiers from its children.
+        :param node: starting node.
+        :return: None.
+        """
+        for child in node.children:
+            if child.type in NODE_TYPES[lang]:
+                start, end = get_positional_bytes(child)
+                identifier = code[start:end].decode("utf-8").lower()
+                if "\n" not in identifier:  # Will break output files. Can add other bad characters later
```

Comment:
I'm still not sure what the issue is. It's not the problem with tree-sitter, it's a problem with topic modeling. I will run it again to find it, but it was encountered when Java-parsing **non-Java file** (test), so for some reason a token had "\n" in it. And when I say it broke our files, I mean topic modeling temporary files, because they are numbered by lines.